compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -730,26 +715,246 @@ fn encode_ty<'tcx>(
     typeid
 }
 
+/// Returns true if the FnAbi is an unpacked "rust-call" ABI closure call.
+fn is_closure_call<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> bool {
+    if fn_abi.conv == Conv::Rust
+        && !fn_abi.c_variadic
+        && !fn_abi.args.is_empty()
+        && let ty::Closure(_, args) = fn_abi.args[0].layout.ty.peel_refs().kind()
+    {
+        let fn_sig = tcx.signature_unclosure(args.as_closure().sig(), hir::Unsafety::Normal);
+        return fn_sig
+            .skip_binder()
+            .inputs()
+            .iter()
+            .eq_by(&fn_abi.args[1..], |ty, arg| *ty == arg.layout.ty)
+            && fn_sig.output().skip_binder() == fn_abi.ret.layout.ty;
+    }
+    false
+}
+
+/// Returns true if the FnAbi is a dynamic Fn trait call.
+fn is_dynamic_fn_call<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> bool {
+    if fn_abi.conv == Conv::Rust
+        && !fn_abi.c_variadic
+        && !fn_abi.args.is_empty()
+        && let ty::Dynamic(predicates, ..) = fn_abi.args[0].layout.ty.peel_refs().kind()
+    {
+        if predicates.is_fn_trait(tcx) {
+            return predicates
+                .fn_trait_args(tcx)
+                .unwrap()
+                .iter()
+                .eq_by(&fn_abi.args[1..], |ty, arg| ty == arg.layout.ty)
+                && predicates.fn_trait_output(tcx).unwrap() == fn_abi.ret.layout.ty;
+        }
+    }
+    false
+}
+
+/// Returns true if the trait object is an Fn trait object.
+fn is_dynamic_fn_trait<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> bool {
```

Comment:
Can you inline this function?

Reply:
Done.

Reply:
Hmm. I may be mistaken but I think that errs meant manually inline it, because it's not clearer to factor out a small single-use predicate.